<a href="https://colab.research.google.com/github/yuanjinren/DS-Unit-2-Linear-Models/blob/master/module2-regression-2/YuanjinRen_LS_DS_212_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Lambda School Data Science

*Unit 2, Sprint 1, Module 2*

---

# Regression 2

## Assignment

You'll continue to **predict how much it costs to rent an apartment in NYC,** using the dataset from renthop.com.

- [ ] Do train/test split. Use data from April & May 2016 to train. Use data from June 2016 to test.
- [ ] Engineer at least two new features. (See below for explanation & ideas.)
- [ ] Fit a linear regression model with at least two features.
- [ ] Get the model's coefficients and intercept.
- [ ] Get regression metrics RMSE, MAE, and $R^2$, for both the train and test data.
- [ ] What's the best test MAE you can get? Share your score and features used with your cohort on Slack!
- [ ] As always, commit your notebook to your fork of the GitHub repo.


#### [Feature Engineering](https://en.wikipedia.org/wiki/Feature_engineering)

> "Some machine learning projects succeed and some fail. What makes the difference? Easily the most important factor is the features used." — Pedro Domingos, ["A Few Useful Things to Know about Machine Learning"](https://homes.cs.washington.edu/~pedrod/papers/cacm12.pdf)

> "Coming up with features is difficult, time-consuming, requires expert knowledge. 'Applied machine learning' is basically feature engineering." — Andrew Ng, [Machine Learning and AI via Brain simulations](https://forum.stanford.edu/events/2011/2011slides/plenary/2011plenaryNg.pdf) 

> Feature engineering is the process of using domain knowledge of the data to create features that make machine learning algorithms work. 

#### Feature Ideas
- Does the apartment have a description?
- How long is the description?
- How many total perks does each apartment have?
- Are cats _or_ dogs allowed?
- Are cats _and_ dogs allowed?
- Total number of rooms (beds + baths)
- Ratio of beds to baths
- What's the neighborhood, based on address or latitude & longitude?

## Stretch Goals
- [ ] If you want more math, skim [_An Introduction to Statistical Learning_](http://faculty.marshall.usc.edu/gareth-james/ISL/ISLR%20Seventh%20Printing.pdf),  Chapter 3.1, Simple Linear Regression, & Chapter 3.2, Multiple Linear Regression
- [ ] If you want more introduction, watch [Brandon Foltz, Statistics 101: Simple Linear Regression](https://www.youtube.com/watch?v=ZkjP5RJLQF4)
(20 minutes, over 1 million views)
- [ ] Add your own stretch goal(s) !

In [0]:
%%capture
import sys

# If you're on Colab:
if 'google.colab' in sys.modules:
    DATA_PATH = 'https://raw.githubusercontent.com/LambdaSchool/DS-Unit-2-Applied-Modeling/master/data/'
    !pip install category_encoders==2.*

# If you're working locally:
else:
    DATA_PATH = '../data/'
    
# Ignore this Numpy warning when using Plotly Express:
# FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
import warnings
warnings.filterwarnings(action='ignore', category=FutureWarning, module='numpy')

In [0]:
import numpy as np
import pandas as pd

# Read New York City apartment rental listing data
df = pd.read_csv(DATA_PATH+'apartments/renthop-nyc.csv')
assert df.shape == (49352, 34)

# Remove the most extreme 1% prices,
# the most extreme .1% latitudes, &
# the most extreme .1% longitudes
df = df[(df['price'] >= np.percentile(df['price'], 0.5)) & 
        (df['price'] <= np.percentile(df['price'], 99.5)) & 
        (df['latitude'] >= np.percentile(df['latitude'], 0.05)) & 
        (df['latitude'] < np.percentile(df['latitude'], 99.95)) &
        (df['longitude'] >= np.percentile(df['longitude'], 0.05)) & 
        (df['longitude'] <= np.percentile(df['longitude'], 99.95))]

#Do train/test split. 
Use data from April & May 2016 to train. Use data from June 2016 to test.

In [0]:
df.head(3)

bathrooms  bedrooms  ... wheelchair_access common_outdoor_space
0        1.5         3  ...                 0                    0
1        1.0         2  ...                 0                    0
2        1.0         1  ...                 0                    0

[3 rows x 34 columns]

In [0]:
df.shape

(48817, 34)

In [0]:
df1=df.copy()    #copy the original dataframe

In [0]:
df1.shape

(48817, 34)

In [0]:
df1['created_dt'] = pd.to_datetime(df1['created'])    # transfer data in 'created' column to datetime datatype 
df1.head()

bathrooms  bedrooms  ... common_outdoor_space          created_dt
0        1.5         3  ...                    0 2016-06-24 07:54:24
1        1.0         2  ...                    0 2016-06-12 12:19:27
2        1.0         1  ...                    0 2016-04-17 03:26:41
3        1.0         1  ...                    0 2016-04-18 02:22:02
4        1.0         4  ...                    0 2016-04-28 01:32:41

[5 rows x 35 columns]

In [0]:
df1['rooms_total'] = df1['bathrooms'] + df1['bedrooms']    # create a new column with numbers of all rooms
df1.head()

bathrooms  bedrooms  ...          created_dt rooms_total
0        1.5         3  ... 2016-06-24 07:54:24         4.5
1        1.0         2  ... 2016-06-12 12:19:27         3.0
2        1.0         1  ... 2016-04-17 03:26:41         2.0
3        1.0         1  ... 2016-04-18 02:22:02         2.0
4        1.0         4  ... 2016-04-28 01:32:41         5.0

[5 rows x 36 columns]

In [0]:
df1['perks_total'] = (df1['balcony'] + df1['cats_allowed'] + df1['common_outdoor_space'] + df1['dining_room']
+ df1['dishwasher'] + df1['dogs_allowed'] + df1['elevator'] + df1['exclusive'] + df1['fitness_center']
+ df1['garden_patio'] + df1['hardwood_floors'] + df1['high_speed_internet'] + df1['loft']
+ df1['outdoor_space'] + df1['pre-war'] + df1['roof_deck'] + df1['terrace'] + df1['wheelchair_access']
+ df1['new_construction'] + df1['swimming_pool'] + df1['laundry_in_building'] + df1['laundry_in_unit']
+ df1['doorman'] + df1['no_fee'])
df1.head()    #create a column with numbers of all perks

bathrooms  bedrooms  ... rooms_total perks_total
0        1.5         3  ...         4.5           0
1        1.0         2  ...         3.0           5
2        1.0         1  ...         2.0           3
3        1.0         1  ...         2.0           2
4        1.0         4  ...         5.0           1

[5 rows x 37 columns]

In [0]:
df1['perks_total'].value_counts()

3     7259
2     5875
4     5766
5     4942
1     4795
6     3853
0     3729
7     2985
8     2459
9     2039
10    1552
11    1221
12     893
13     659
14     393
15     214
16     107
17      60
18      12
19       4
Name: perks_total, dtype: int64

In [0]:
df1['created_dt'].dt.month

0        6
1        6
2        4
3        4
4        4
        ..
49347    6
49348    4
49349    4
49350    4
49351    4
Name: created_dt, Length: 48817, dtype: int64

In [0]:
train = df1[(df1['created_dt'].dt.month == 4) |(df1['created_dt'].dt.month == 5)]   #get train set
train

bathrooms  bedrooms  ... rooms_total perks_total
2            1.0         1  ...         2.0           3
3            1.0         1  ...         2.0           2
4            1.0         4  ...         5.0           1
5            2.0         4  ...         6.0           0
6            1.0         2  ...         3.0           3
...          ...       ...  ...         ...         ...
49346        1.0         1  ...         2.0           5
49348        1.0         1  ...         2.0           9
49349        1.0         1  ...         2.0           5
49350        1.0         0  ...         1.0           5
49351        1.0         2  ...         3.0           1

[31844 rows x 37 columns]

In [0]:
test = df1[df1['created_dt'].dt.month == 6]    #get test set
test

bathrooms  bedrooms  ... rooms_total perks_total
0            1.5         3  ...         4.5           0
1            1.0         2  ...         3.0           5
11           1.0         1  ...         2.0           0
14           1.0         1  ...         2.0           3
24           2.0         4  ...         6.0          11
...          ...       ...  ...         ...         ...
49305        1.0         2  ...         3.0           2
49310        1.0         3  ...         4.0           8
49320        1.0         1  ...         2.0           2
49332        1.0         2  ...         3.0           4
49347        1.0         2  ...         3.0           5

[16973 rows x 37 columns]

# Begin with baselines (0 features)

In [0]:
#Begin with baseline
train['price'].mean()

3575.604007034292

In [0]:
#Arrange y target vectors
target = 'price'
y_train = train[target]
y_test = test[target]

In [0]:
# Get mean baseline
guess = y_train.mean()
guess

3575.604007034292

In [0]:
#train error
from sklearn.metrics import mean_absolute_error
y_pred = [guess] * len(y_train)
mae = mean_absolute_error(y_train, y_pred)
print(f'Train error: {mae:.2f}')

Train error: 1201.88


In [0]:
#test error
y_pred = [guess] * len(y_test)
mae = mean_absolute_error(y_test,y_pred)
print(f'Test error: {mae:.2f}')

Test error: 1197.71


# Regression (2 features) 

In [0]:
# Visualize relationshp between 2 features and the price (target)
import plotly.express as px
px.scatter_3d(
    train,
    x = 'rooms_total',
    y = 'perks_total',
    z = 'price',
    title = 'Costs to Rent an Apartment in NYC, 2016/04-2016/05'
)

In [0]:
# Import the appropriate class from Scikit-learn
from sklearn.linear_model import LinearRegression

In [0]:
# Instantiate this class
model = LinearRegression()

In [0]:
# re-arrange features matrices
features = ['rooms_total','perks_total']
print(f'Linear Regression, dependent on: {features}')

Linear Regression, dependent on: ['rooms_total', 'perks_total']


In [0]:
X_train = train[features]
X_test = test[features]
print(X_train.shape)
print(X_test.shape)

(31844, 2)
(16973, 2)


In [0]:
#fit the model
model.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [0]:
# apply the model to new data
y_pred_train = model.predict(X_train)
y_pred_train

array([2839.4009342 , 2737.63658265, 4939.74209926, ..., 3042.92963729,
       2274.97301457, 3403.82885382])

In [0]:
y_pred_train.shape

(31844,)

In [0]:
# calculate MAE of train set
mae_train = mean_absolute_error(y_train, y_pred_train)
print(f'MAE of train set is: {mae_train:.2f}')

MAE of train set is: 856.62


In [0]:
from sklearn.metrics import mean_squared_error

In [0]:
# calculate MSE of train set
mse_train = mean_squared_error(y_train,y_pred_train)
mse_train

1678877.2727735792

In [0]:
# calculate RMSE of train set
rmse_train = np.sqrt(mse_train)
rmse_train

1295.7149658677172

In [0]:
from sklearn.metrics import r2_score

In [0]:
# calculate R squared of train set
r2_train = r2_score(y_train, y_pred_train)
r2_train

0.45930369872329946

It means that 46% of the price variance can be explained by the two features including 'rooms_total' and 'perks_total' for train set. 

In [0]:
y_pred_test = model.predict(X_test)
y_pred_test

array([4453.99943635, 3810.88626001, 2534.10787955, ..., 2737.63658265,
       3709.12190846, 3810.88626001])

In [0]:
# calculate MAE of test set
mae_test = mean_absolute_error(y_test,y_pred_test)
print(f'MAE of test set is: {mae_test:.2f}')

MAE of test set is: 868.00


In [0]:
# calculate MSE of test set
mse_test = mean_squared_error(y_train,y_pred_train)
mse_test

1678877.2727735792

In [0]:
# calculate RMSE of test set
rmse_test=np.sqrt(mse_test)
rmse_test

1295.7149658677172

In [0]:
# calculate R Squared of test set
r2_test = r2_score(y_test,y_pred_test)
r2_test

0.46189791795456303

It means that 46.2% of the price variance can be explained by the two features including 'rooms_total' and 'perks_total' for test set. 

# Intercept and Coefficients

In [0]:
model.intercept_,model.coef_

(998.1946341143803, array([767.95662272, 101.76435155]))

In [0]:
beta0=model.intercept_
beta1, beta2 = model.coef_
print(f'y={beta0:.2f} + {beta1:.2f}*x1 + {beta2:.2f}*x2')

y=998.19 + 767.96*x1 + 101.76*x2


In [0]:
print('Intercept for the above linear regression model is: ', model.intercept_)
coefficients = pd.Series(model.coef_,features)
print(coefficients.to_string())

Intercept for the above linear regression model is:  998.1946341143803
rooms_total    767.956623
perks_total    101.764352


Since the coefficients are both positive, two features are positive related to price of renting an apartment. 

# Plot the plane of best fit

In [0]:
import itertools
import numpy as np
import plotly.express as px
import plotly.graph_objs as go
from sklearn.linear_model import LinearRegression

In [0]:
def regression_3d (df, x, y, z, num = 100, **kwargs):
  fig = px.scatter_3d(df,x,y,z,**kwargs)
  features = [x,y]
  target = z
  model = LinearRegression()
  model.fit(df[features], df[target])

  xmin,xmax = df[x].min(), df[x].max()
  ymin,ymax = df[y].min(),df[y].max()

  xcoords = np.linspace(xmin, xmax, num)
  ycoords = np.linspace(ymin, ymax, num)
  coords = list(itertools.product(xcoords, ycoords))

  predictions = model.predict(coords)
  Z = predictions.reshape(num,num).T

  fig.add_trace(go.Surface(x=xcoords,y=ycoords,z=Z))
  return fig

In [0]:
regression_3d(train,
              x = 'rooms_total',
              y = 'perks_total',
              z = 'price',
              title = 'Costs to Rent an Apartment in NYC, 2016/04-2016/05')